internal/testclient/testtx/context.go

```diff
@@ -0,0 +1,350 @@
+func NewOneTimeErrTxTimeoutTxContext(
+	t *testing.T,
+	keyring cosmoskeyring.Keyring,
+	signingKeyName string,
+	expectedTx *cometbytes.HexBytes,
+	expectedTxHash *cometbytes.HexBytes,
+	expectedErrMsg *string,
+) *mockclient.MockTxContext {
+	t.Helper()
+
+	signerKey, err := keyring.Key(signingKeyName)
+	require.NoError(t, err)
+
+	signerAddr, err := signerKey.GetAddress()
+	require.NoError(t, err)
+
+	*expectedErrMsg = fmt.Sprintf(
+		"fee payer address: %s does not exist: unknown address",
+		signerAddr.String(),
+	)
+
+	txCtxMock := NewBaseTxContext(
+		t, signingKeyName,
+		keyring,
+		expectedTx,
+		expectedTxHash,
+	)
+
+	// intercept #BroadcastTxSync() call to mock response and prevent actual broadcast
+	txCtxMock.EXPECT().BroadcastTxSync(gomock.Any()).
+		DoAndReturn(func(txBytes []byte) (*cosmostypes.TxResponse, error) {
+			return &cosmostypes.TxResponse{
+				Height:    1,
+				TxHash:    expectedTxHash.String(),
+				RawLog:    "",
+				Logs:      nil,
+				Tx:        nil,
+				Timestamp: "",
+				Events:    nil,
+			}, nil
+		}).Times(1)
+
+	txCtxMock.EXPECT().QueryTx(
+		gomock.AssignableToTypeOf(context.Background()),
+		gomock.AssignableToTypeOf([]byte{}),
+		gomock.AssignableToTypeOf(false),
+	).DoAndReturn(func(
+		ctx context.Context,
+		txHash []byte,
+		_ bool,
+	) (*cometrpctypes.ResultTx, error) {
+		return &cometrpctypes.ResultTx{
+			Hash:   txHash,
+			Height: 1,
+			TxResult: abci.ResponseDeliverTx{
+				Code:      1,
+				Log:       *expectedErrMsg,
+				Codespace: "test_codespace",
+			},
+			Tx: expectedTx.Bytes(),
+		}, nil
+	})
+
+	return txCtxMock
+}
+
+// NewOneTimeErrCheckTxTxContext creates a mock transaction context to simulate
+// a specific error scenario during the ABCI check-tx phase (i.e., during initial
+// validation before the transaction is included in the block).
+//
+// Parameters:
+// - t: The testing.T instance for the current test.
+// - keyring: The Cosmos SDK keyring containing the signer's cryptographic keys.
+// - signingKeyName: The name of the key within the keyring to be used for signing.
+// - expectedTx: A pointer whose value will be set to the expected transaction
+// bytes (in hexadecimal format).
+// - expectedTxHash: A pointer whose value will be set to the expected
+// transaction hash.
+// - expectedErrMsg: A pointer whose value will be set to the expected error
+// message string.
+//
+// The function operates as follows:
+//  1. Retrieves the signer's cryptographic key from the provided keyring based on
+//     the signingKeyName.
+//  2. Determines the corresponding address of the signer's key.
+//  3. Composes an error message suggesting that the fee payer's address is unrecognized.
+//  4. Creates a base mock transaction context using the NewBaseTxContext function.
+//  5. Sets up the mock behavior for the BroadcastTxSync method to return a specific
+//     error response related to the check phase of the transaction.
+func NewOneTimeErrCheckTxTxContext(
```

Comment:
Question: Why not have the `Times(1)` be `Times(N)`, remove `OneTime` from the function name and provide `N` as a parameter.

Not a strong preference but curious to hear what you think.

Reply:
That's a reasonable suggestion IMO. I've mostly been letting the tests drive the shape of the helpers. We don't have  any tests which need to produce an error like this more than once. As it's a pretty straightforward refactor in this case, I would prefer to defer it to when the need arises.

Reply:
**TL;DR: will note in a comment but deferring for now.**

I think that could work but we should be careful about how we set that precedent. There can be multiple mock objects to call `Times()` on which may not necessarily have a constant correlation. In some cases we may even prefer to call `MinTimes()` or `MaxTimes()`,  in which case, it gets more  complicated. I think the unambiguous version of this would end up looking something like:

```go
package  mockthing

func NewThing(t *testing.T, startTimes, doSomethingTimes, repeatMinTimes int) {
  thingMock :=  mockpkg.NewMockThing()
  thingMock.EXPECT().Start(...).Times(startTimes)
  thingMock.EXPECT().DoSomething(...).Times(doSomethingTimes)
  thingMock.EXPECT().Repeat(...).MinTimes(repreatMinTimes)
}
```

Which is probably fine, but unless there's a need for it,  I think I would prefer to wait until such a time  as there is to do that refactor.

Reply:
Makes sense to me!